### ISSUE TITLE
[META] Refactoring the server monolith

### ISSUE THREAD

**Issue by nknize** [opensearch-project/OpenSearch#8110]
Related discuss issue: #5910 

[WIP] This meta issue tracks the scope for refactoring the monolithic `:server` module into separable libraries to support JPMS and serverless or cloud native implementation extensions.

Phase 0 - JPMS Support (Eliminate top level split packages)
- [ ] o.o.common split package - refactor classes to `:libs:opensearch-common` or `:libs:opensearch-core` where needed
- [ ] o.o.bootstrap split package - refactor to o.o.bootstrap in `:libs:opensearch-common` to new `o.o.common.bootstrap` package
- [ ] o.apache.lucene split package - refactor to o.o.core.lucene in `:libs:opensearch-core` or o.o.lucene in `:libs:opensearch-server`
- [ ] o.o.cli split package - refactor to o.o.core.cli in `:libs:opensearch-core` or o.o.cli in `:libs:opensearch-cli`

Phase 1 - Decoupling to support serverless and cloud native implementation extensions
 
- [ ] o.o.env - refactor classes to o.o.core.env in `:libs:opensearch-core` (dependency of o.o.bootstrap)
- [ ] o.o.common.settings - refactor classes to o.o.core.common.settings in `:libs:opensearch-core` (dependency of o.o.env)
- [ ] o.o.cluster.metadata - refactor select classes to a new `:libs:opensearch-cluster` library

Resulting libraries:

`:libs:opensearch-common`
`:libs:opensearch-core`
`:libs:opensearch-cluster`

**Comment by nknize** [comment]
/cc @rishabh6788 this work has been largely advancing by me over the last few months and reaching some good breakthrough milestones. The Streamable classes (e.g., StreamInput, StreamOutput, Writeable, NamedWriteable) and OpenSearchException class [PR](https://github.com/opensearch-project/OpenSearch/pull/8035) is up to tease out a big part of the tightly coupled classes across the `:server` module into separable libraries. This unlocks a big portion of `:server`. The meta issue here seeks to resolve the remaining split packages across server as outlined in the #1838 issue you opened a while ago. A separate goal of #5910 is to ensure as many of the existing extensible classes are in the right library and right package name such that we can thoughtfully expose or restrict extension points to downstream plugin consumers. 

If you have some time I'd love for you to jump in and help review the PRs.

**Comment by andrross** [comment]
Thanks @nknize! I want to start the discussion on expediting the completion of phase 0 (elimination of packages split across jars) AND backporting it the 2.x branch. I think the current state where there is major structural divergence between `main` and `2.x` is undesirable due to causing conflicts on backports. Given that plugins also maintain a branch developing against OpenSearch `main`, I think they are dealing with the same problem. Also, plugins are not binary-compatible across versions anyway: they must at a minimum update version numbers and re-compile. I think it easier for everyone to complete the structural refactoring quickly and do it across both branches (but also not do it near a release to not add any extra variables as teams are completing integration for that release).

@reta @dblock What do you think?

**Comment by reta** [comment]
> I think it easier for everyone to complete the structural refactoring quickly and do it across both branches (but also not do it near a release to not add any extra variables as teams are completing integration for that release).

@andrross  Agree :100: , the elimination of packages splits would be a huge win towards getting JPMS done, though for `2.x` this is going to be 100% breaking change (that goes way beyond `@opensearch.internal`). Would we selectively keep the structural differences between `2.x` and `3.x`+ ?

**Comment by nknize** [comment]
> Would we selectively keep the structural differences between 2.x and 3.x+ ?

I think we should look at that on a case by case basis? Let's lean into progress not perfection here so we can iterate quickly. 

I also think we shouldn't have FUD around breaking java signatures and package names in minor releases at this juncture (as long as we don't do it too close to a release). The codebase isn't well defined enough for that (hence `@opensearch.internal` marked on the majority of the codebase). Once we have JPMS and we start defining exports in `module-info.java` we can use that as the mechanism to collectively invent and simplify on the java surface area we want to expose external to core and guarantee non-breaking across minors on those classes.

**Comment by andrross** [comment]
> I think we should look at that on a case by case basis?

I agree with a case by case basis. I think the value here of getting rid of a blocker to adopting JPMS is clear, and the "breakages" will be super simple to resolve as it is almost exclusively a matter of fixing imports (and the plugins maintaining a version against `main` have to resolve these issues anyway), so there's a compelling case for doing the items marked as phase 0 now.

I also agree that it is not practical to be 100% strict on not changing any Java APIs given how tightly coupled the code bases are.

**Comment by nknize** [comment]
Semi related, I [opened an issue](https://github.com/opensearch-project/OpenSearch/issues/8127) for someone to add useful Java annoations we could use to enforce bwc. Not a requirement for this but another mechanism that would be helpful once we start carving out the publicly exposed java classes through `module-info.java`.

**Comment by dblock** [comment]
> @reta @dblock What do you think?

The extensibility project (#2447) proposed that we first make plugins depend on an SDK reducing the N plugins : 1 core dependency tree to N plugins : 1 SDK : 1 core, first, therefore avoiding this pain (only the SDK will need to change when core changes), and we are at experimental release for 2.9, but it looks like we're not choosing to wait for that to deliver and aren't going to try to port plugins to extensions. I don't have any solutions or workarounds to what the downstream dependencies are experiencing or will experience, you just can't both refactor and not cause disruption.

Given that, I would keep doing what @nknize is doing, and GA 3.0 as soon as possible to stop the pain of backports to 2.x.

**Comment by rishabh6788** [comment]
> /cc @rishabh6788 this work has been largely advancing by me over the last few months and reaching some good breakthrough milestones. The Streamable classes (e.g., StreamInput, StreamOutput, Writeable, NamedWriteable) and OpenSearchException class [PR](https://github.com/opensearch-project/OpenSearch/pull/8035) is up to tease out a big part of the tightly coupled classes across the `:server` module into separable libraries. This unlocks a big portion of `:server`. The meta issue here seeks to resolve the remaining split packages across server as outlined in the #1838 issue you opened a while ago. A separate goal of #5910 is to ensure as many of the existing extensible classes are in the right library and right package name such that we can thoughtfully expose or restrict extension points to downstream plugin consumers.
> 
> If you have some time I'd love for you to jump in and help review the PRs.

Tagging correct Rishabh @rishabhmaurya

**Comment by nknize** [comment]
> ...GA 3.0 as soon as possible to stop the pain of backports to 2.x.

+1 to work towards 3.0 GA


Backport pain will always exist
Within a repo when a namespace changes. 


> ...first make plugins depend on an SDK reducing the N plugins : 1 core dependency tree to N plugins : 1 SDK : 1 core, first, therefore avoiding this pain

A big gap in this plan is the sdk still [transitively exports](https://github.com/opensearch-project/opensearch-sdk-java/blob/72cc96231f9265cce5583b6b29e381a664f29935/build.gradle#L145) all of the core classes to "downstream" plugins defeating strong encapsulation provided by JPMS (which is the primary objective of phase 0). This means core isolation through the sdk is not achieved and plugins can bypass the sdk and override core logic. This is exacerbated by the sdk also itself [not providing JPMS](https://github.com/opensearch-project/opensearch-sdk-java/tree/main/src/main/java/org/opensearch/sdk/api) support. Opening RFCs to deprecate core mechanisms [like ThreadContext](https://github.com/opensearch-project/security/issues/2860) is a side effect we can eliminate with JPMS support.  The near term pain (reduced by good DevOps mechanisms like `checkCompatibility`) I think is worth it to achieve the strong encapsulation we need to make SDK truly isolate the core classes. 

These efforts are far better together.

**Comment by andrross** [comment]
> GA 3.0 as soon as possible to stop the pain of backports to 2.x

Major versions upgrades are disruptive to all users (including non-plugin developer users). I have a hard time justifying the disruption to the community at large that a major version release brings with the refactoring changes here.

> I don't have any solutions or workarounds to what the downstream dependencies are experiencing or will experience, you just can't both refactor and not cause disruption.

Assuming that downstream dependencies are maintaining a version against main and against 2.x, then they are already experiencing the disruption, and the way to resolve that disruption is to backport the refactorings and bring the branches in line with each other. Any community plugins (outside of the opensearch-project) that are only maintaining a version against the 2.x branch are blissfully unaware of these structural changes and would be disrupted by the backports. But we don't truly support semver with plugins today anyway because plugins have to make code changes and recompile for all releases. We should still be thoughtful about the "breaking" changes that we backport, but I'm trying to make the case that the changes in phase 0 here (structural that are trivial for a consumer to resolve) are worth it on balance.

**Comment by nknize** [comment]
Once Phase 0 is complete we can selectively `export` specific packages we want downstreams to be able to import and extend. Those we don't can be refactored to an `.internal` namespace that's not exposed. Technically this completely removes the need for OpenSearch-SDK because the SDK is already defined and implemented through the core (libraries and modules). To make life easier we could just move those exported API classes to new `.sdk.` sub packages in the existing libraries.  This enables us to finally provide enforceable API BWC and the breaking changes will be all but gone (not going to say completely eliminated because there may be some corner cases). This will also enable us begin working on actually supporting semver in the public API.  It greatly simplifies the extensibility implementation  because it reduces it further from the `N plugins : 1 SDK : 1 core` to `N plugins : 1 core w/ "SDK"`

**Comment by dblock** [comment]
> A big gap in this plan is the sdk still [transitively exports](https://github.com/opensearch-project/opensearch-sdk-java/blob/72cc96231f9265cce5583b6b29e381a664f29935/build.gradle#L145) all of the core classes to "downstream" plugins defeating strong encapsulation provided by JPMS (which is the primary objective of phase 0). 

This is only current implementation for build time, not at runtime, to go faster. At runtime extensions do not depend on anything transitively.

**Comment by dblock** [comment]
> Once Phase 0 is complete we can selectively `export` specific packages we want downstreams to be able to import and extend. Those we don't can be refactored to an `.internal` namespace that's not exposed. Technically this completely removes the need for OpenSearch-SDK because the SDK is already defined and implemented through the core (libraries and modules). 

This is misleading. You still have jar hell. The SDK doesn't just move classes around or annotate them, it introduces a proper r/w API, and allows to have compatibility across multiple (major) versions, and to fully remote extensions on a separate runtime that is not core.

**Comment by nknize** [comment]
> This is only current implementation for build time...

At build time is _exactly_ where all of core classes are exposed such that a developer can override core logic.

> At runtime extensions do not depend on core transitively.

At runtime the classloader needs core classes directly (e.g., ClusterState for [SDKClusterService](https://github.com/opensearch-project/opensearch-sdk-java/blob/72cc96231f9265cce5583b6b29e381a664f29935/src/main/java/org/opensearch/sdk/SDKClusterService.java#L57)) without the runtime dependency you'd get a CNF exception. This is what modularity prevents.

**Comment by dblock** [comment]
The goals of refactoring is to avoid leaking internals into plugins by establishing a r/w API with strong encapsulation. Even when achieved, it will be difficult to have a large ecosystem of plugins for all the reasons that are not solved by refactoring. 

If plugins don't depend on core at all, then this benefit is not needed, and thus refactoring unnecessary for that purpose (it has other benefits). 

I think we should do both, but I believe that extensions give a lot more benefits a lot quicker, especially that plugin authors don't have to work or deal with core anymore, and don't need to release with every new minor version of OpenSearch, making developing plugins a lot cheaper.

**Comment by dbwiddis** [comment]
> If ~plugins~ extensions don't depend on core at all, then this benefit is not needed, and thus refactoring unnecessary for that purpose (it has other benefits).

@dblock In an ideal future world (thinking 3.x) this is the goal for extensions. The only commonality is intended to be request/response classes auto-generated from spec files... and probably some needed XContent parsing classes.

**Comment by dblock** [comment]
> Even with opensearch-sdk the ~plugins~ extensions depend on core.

That is simply incorrect.

Re: value of extensions I invite you to read https://github.com/opensearch-project/opensearch-sdk-java#introduction.

**Comment by nknize** [comment]
> That is simply incorrect.

AnomalyDetection takes a [dependency on core](https://github.com/opensearch-project/anomaly-detection/blob/aedb781a6b1df984636d72d50b65d41138539918/build.gradle#L115). At least it's `implementation` dependency is (marginally) better than opensearch-sdk's [`api` dependency](https://github.com/opensearch-project/opensearch-sdk-java/blob/54832bc3e3adf9bd43cfd51e0f88097bc7502120/build.gradle#L145). To load the core classes needed at runtime means the classes have to be available to the class loader.. hence, "dependency".  Let's take another example. `AnalysisExtension` which has a [slew of core dependency](https://github.com/opensearch-project/opensearch-sdk-java/blob/main/src/main/java/org/opensearch/sdk/api/AnalysisExtension.java#L17) classes, many which [rely on other core classes like `o.o.env.Environment`](https://github.com/opensearch-project/OpenSearch/blob/68c1c86897a59fcffa77bfecc06641326e5e5199/server/src/main/java/org/opensearch/index/analysis/PreBuiltAnalyzerProviderFactory.java#L40)


> I invite you to read...

I've read it...now we're in discussion mode. The part I've always liked about extensibility is `"our goal is to isolate plugin interactions with OpenSearch"`.  This is achieved by running in a separate `DiscoveryNode`, but let's be clear, this is starting up a new node. So it's being achieved through isolating via another node. That's not a bad thing (though it takes some bootstrapping). I just want to ensure we don't oversell extensibility as solving the "sdk" isolation conversation. It's still taking dependencies on core. I prefer we handle this isolation in core itself and keep the "sdk" api components in the core, not in `opensearch-sdk`.

**Comment by nknize** [comment]
> comment above explains that this dependency is temporary..

That's in regard to ClusterState, I'm talking about the project as a whole (e.g., "Let's take another example. AnalysisExtension..."). You're not going to get away from requiring opensearch-sdk transitively taking a dependency on both core and lucene. Analyzers implement lucene Tokenizers, TokenStream, etc. What's the point in attempting to write all sorts of opensearch-sdk scaffolding logic to try and wrap these when JPMS already solves your encapsulation? So let's not oversell that extensions is going to provide something it's not going to provide. I think the real value in extensions is isolating the plugin such that it doesn't crash your entire cluster. THAT is what the mission of extensibility should be. Not "eliminate the need to take a dependency on core".

**Comment by reta** [comment]
@nknize (not adding oil but sharing some thoughts)  my understanding of the extensions is very much aligned with what @dblock has been saying. To me, as the end result is that Java extensions do no depend on core but sdk (and sdk consequently should not have any dependency of core), otherwise we are repeating the plugins path pretty much:
 - the vision for extension (at large) was that we could support not only Java but other languages
 - the extensions would not be able to completely replace plugins but provide alternative means for what plugins used to be the only answer
 - the extensions could depend on Apache Lucene (directly, not transitively)

But ...

> Let's take another example. AnalysisExtension which has a [slew of core dependency](https://github.com/opensearch-project/opensearch-sdk-java/blob/main/src/main/java/org/opensearch/sdk/api/AnalysisExtension.java#L17) classes, many which [rely on other core classes like o.o.env.Environment](https://github.com/opensearch-project/OpenSearch/blob/68c1c86897a59fcffa77bfecc06641326e5e5199/server/src/main/java/org/opensearch/index/analysis/PreBuiltAnalyzerProviderFactory.java#L40)

... completely changes the extension model for me, replicating this feature (and many similar) without depending on core would be a nightmare (I don't see any other viable choices here vs depending on core). 

> What's the point in attempting to write all sorts of opensearch-sdk scaffolding logic to try and wrap these when JPMS already solves your encapsulation? 

We are asking a bit to much from JPMS here, extension author may decide they don't want to use JPMS but rely on classpath instead, that would lift the restrictions (this is how we use modular Apache Lucene today essentially). I don't know how could we enforce that (may be variation JarHell could be an answer actually).

**Comment by nknize** [comment]
> ...the end result is that Java extensions do no depend on core but sdk (and sdk consequently should not have any dependency of core),


Maybe my message is being missed but I'm not saying anything different from a theoretical pov. Of course it would be nice for an extension / plugin to have a one line dependency (call it opensearch-sdk, opensearch-api it doesn't matter and how it's packaged is separate from implementation). What I'm saying is to look at reality, and I'm giving very real use cases where the extensibility "SDK" proposal does not work. 

Let me be direct. Extensibility should be separate from the SDK (e.g., it uses and interacts with the SDK but does not define it). Leave the "extensibility" conversation to talking about the ExtensionsManager and how the distributed system (firing up extensions logic) launches the extension in its own discovery node.  SDK is a different problem and JPMS in the core repo (libs, modules, plugins) plays an integral part here the current opensearch-sdk design ignores.  


> replicating this feature (and many similar) without depending on core would be a nightmare

Yep. Same with aggs, scripted aggs, query languages, and many compute scenarios (e.g., spark SQL). This is the reality we live in. So I'm asking we be pragmatic not wishful and theoretical. (e.g., quantum physics says I can theoretically run through a wall unscathed).

> We are asking a bit to much from JPMS here, extension author may decide they don't want to use JPMS .

I'm not talking about JPMS in an extension. I'm talking about JPMS for strong encapsulation in the core repo classes. To not leak the core classes and expose vulnerabilities downstreams (e.g., security) is already looking at immutability changes that would be unnecessary if JPMS were already being used. Let's stop the "reinvention" of mechanisms when many goals are already solved by the language (like JPMS).

Key takeaways: separate the extensions and sdk conversations. Start the SDK conversation in the core by defining the core API through JPMS module-info.java, not a downstream repo. Anything not enabled by JPMS is addressed on a case by case basis.

**Comment by rishabhmaurya** [comment]
We probably need to address lucene split packages in server module - https://github.com/opensearch-project/OpenSearch/tree/main/server/src/main/java/org/apache/lucene as it would start failing when lucene jars are present in module path?

**Comment by dblock** [comment]
> Start the SDK conversation in the core by defining the core API through JPMS module-info.java, not a downstream repo.

Nomenclature-wise we have `opensearch-core` so maybe calling it "SDK" is more confusing than anything because we created https://github.com/opensearch-project/opensearch-sdk-java and called it the "OpenSearch SDK for Java"?

**Comment by prudhvigodithi** [comment]
Just went through the issue and some points discussed across comments from this issue, part of https://github.com/opensearch-project/OpenSearch/issues/1588 and https://github.com/opensearch-project/OpenSearch/issues/5910. Achieving JPMS is bigger goal and is a bigger effort. We need to start creating modules and plan out for circular dependencies, decide which parts to modularize first and do it in Incremental steps by even making Gradle changes and adding module-info.java files.

For phase-0 (targeting v3.0.0) we can consider eliminating top level split packages as JPMS requires that each package name be uniquely owned by a single module if those modules are on the module path. 

I will start doing some analysis to rename or refactor some packages that avoids split-package conflict when we start with JPMS effort.

**Comment by prudhvigodithi** [comment]
I have updated the phase-0 and phase-1 list with some more refactoring requirement. Still on track to complete phase-0 which is targeted for v3.0.0 release. 

While refactoring `org.apache.lucene` (here is the related PR https://github.com/opensearch-project/OpenSearch/pull/17241) the left over classes from `org.apache.lucene.codecs` and `org.apache.lucene.index` are added for Star Tree implementation, adding @expani to see if this can be removed or if other dependencies can be fetched from Lucene, adding them to OpenSearch and changing the package group from `org.apache.lucene`.

Thanks
@reta @andrross @msfroh @getsaurabh02

**Comment by expani** [comment]
Thanks @prudhvigodithi 
@bharath-techie please track this in a separate issue and let's discuss on the path forward.
